app/src/main/java/com/github/onlynotesswent/utils/TextExtractor.kt

```diff
@@ -0,0 +1,102 @@
+package com.github.onlynotesswent.utils
+
+import android.graphics.Bitmap
+import android.graphics.pdf.PdfRenderer
+import android.os.ParcelFileDescriptor
+import android.util.Log
+import android.widget.Toast
+import androidx.activity.ComponentActivity
+import com.google.mlkit.vision.common.InputImage
+import com.google.mlkit.vision.text.TextRecognition
+import com.google.mlkit.vision.text.TextRecognizer
+import com.google.mlkit.vision.text.latin.TextRecognizerOptions
+import java.io.File
+
+/**
+ * TextExtractor class that detects latin-based text from a PDF file. The class uses the Google ML
+ * Kit Text Recognition API to extract text from the PDF file.
+ *
+ * @param activity the ComponentActivity that will use the TextExtractor
+ */
+class TextExtractor(private val activity: ComponentActivity) {
+  private lateinit var textRecognizer: TextRecognizer
+
+  fun init() {
+    textRecognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)
+  }
+
+  /**
+   * Processes the given PDF file to extract text from it.
+   *
+   * @param pdfFile the PDF file to extract text from.
+   * @param onSuccess the lambda function to call with the extracted text.
+   */
+  fun processPdfFile(pdfFile: File, onSuccess: (String) -> Unit) {
+    if (!::textRecognizer.isInitialized) {
+      // Case should not happen if class is correctly initialized
+      Log.e(TAG, "TextRecognizer is not initialized. Call init() before processing PDF file.")
+      return
+    }
+
+    try {
+      val bitmaps = convertPdfToBitmap(pdfFile)

```

Comment:
Perhaps you could consider offloading this function to a background thread or coroutine in order to not block the main thread for large pdf? However the issue with such solution would be that it can become very hard to test on the CI, as when I tried using asynchronous operations and coroutines the CI crashed. We should discuss this with the team, but for now we can leave this as is.